File: hot/leetcode/8. 字符串转换整数 (atoi)/8. 字符串转换整数 (atoi).go
package main

import "math"

func myAtoi(s string) int {
	res := 0
	sign := 1
	start := 0
	// 去掉前段空格
	n := len(s)
	for start < n && s[start] == ' ' {
		start += 1
	}
	if start == n {
		return 0
	}

	// 存储正负号
	if s[start] == '-' {
		sign = -1
		start += 1
	} else if s[start] == '+' {
		start += 1
	}
	// 把字符串转换为数字
	for start < n && s[start] >= '0' && s[start] <= '9' {
		res = res*10 + sign*int(s[start]-'0')
		start += 1
        // 溢出判断
		if res > math.MaxInt32 {
			return math.MaxInt32
		}
		if res < math.MinInt32 {
			return math.MinInt32
		}
	}

	return res
}

func main() {
	s := "9223372036854775808"
	println(myAtoi(s))
}

File: hot/leetcode/8. 字符串转换整数 (atoi)/8. 字符串转换整数 (atoi).py
"""
1.去掉前段空格
2.判断正负号
3.连续把字符串专为数字
4.溢出判断
"""


class Solution:
    def myAtoi(self, s: str) -> int:
        MAX_INT = 2 ** 31 - 1
        MIN_INT = -2 ** 31
        res = 0
        sign = 1
        start = 0
        # 去掉前段空格
        while start < len(s) and s[start] == ' ':
            start += 1
        if start == len(s):
            return 0
        # 判断正负号
        if s[start] == '-':
            sign = -1
            start += 1
        elif s[start] == '+':
            start += 1
        # 连续把字符串专为数字，并*正负号
        while start < len(s) and s[start].isdigit():
            res = res * 10 + int(s[start])
            start += 1
        res *= sign
        # 溢出判断
        if res > MAX_INT:
            return MAX_INT
        if res < MIN_INT:
            return MIN_INT
        return res


if __name__ == '__main__':
    s = "42"
    print(Solution().myAtoi(s))
